diff --git a/hw/xfree86/common/xf86Init.c b/hw/xfree86/common/xf86Init.c
index 017dcb6..4808f04 100644
--- a/hw/xfree86/common/xf86Init.c
+++ b/hw/xfree86/common/xf86Init.c
@@ -1177,12 +1177,12 @@ ddxProcessArgument(int argc, char **argv, int i)
     }
 
     /* First the options that are not allowed with elevated privileges */
-    if (!strcmp(argv[i], "-modulepath") || !strcmp(argv[i], "-logfile")) {
+    if (!strcmp(argv[i], "-modulepath")) {
         if (xf86PrivsElevated()) {
             FatalError("The '%s' option cannot be used with "
                        "elevated privileges.\n", argv[i]);
         }
-        else if (!strcmp(argv[i], "-modulepath")) {
+        else {
             char *mp;
 
             CHECK_FOR_REQUIRED_ARGUMENT();
@@ -1193,7 +1193,8 @@ ddxProcessArgument(int argc, char **argv, int i)
             xf86ModPathFrom = X_CMDLINE;
             return 2;
         }
-        else if (!strcmp(argv[i], "-logfile")) {
+    }
+    if (!strcmp(argv[i], "-logfile")) {
             char *lf;
 
             CHECK_FOR_REQUIRED_ARGUMENT();
@@ -1203,7 +1204,14 @@ ddxProcessArgument(int argc, char **argv, int i)
             xf86LogFile = lf;
             xf86LogFileFrom = X_CMDLINE;
             return 2;
-        }
+    }
+    if (!strcmp(argv[i], "-nohwaccess"))
+    {
+        return 1; // Root access is REQUIRED
+    }
+    if (!strcmp(argv[i], "-maxvt"))
+    {
+        return 2; // ignore that.
     }
     if (!strcmp(argv[i], "-config") || !strcmp(argv[i], "-xf86config")) {
         CHECK_FOR_REQUIRED_ARGUMENT();
